lib/spack/spack/variant.py

```diff
@@ -302,7 +302,9 @@ def satisfies(self, other):
         """
         # If names are different then `self` does not satisfy `other`
         # (`foo=bar` will never satisfy `baz=bar`)
-        return other.name == self.name
+        # Otherwise we want all the values in `other` to be also in `self`
```

Comment:
The docstring is no longer correct for this function. Is it possible that this could be `NotImplemented` and this definition could be pushed down to `MultiValuedVariant`? Actually it looks like that's how it was initially and then that definition was pulled up to here - why? Is there another subclass of variant which benefits from this implementation?

Reply:
The reason for this is that in the `matrix` operation in Spack stacks, we need to be able to check whether the abstract variant satisfies some other variant.

https://github.com/spack/spack/issues/16841 has an example of where Spack currently fails because the check for abstract variants is too loose.

AbstractVariants are at some point going to become variants of a concrete type (MultiValue, SingleValue, or Boolean). What this does is adopt the most permissive definition of `satisfies` of any variant type, for Abstract variants.

The problem with that, is that abstract variants can be constrained in ways that really don't make sense as "satisfaction". So we needed another way of checking that. That's what `compatible` is doing.